[fix] 모임방 참여 api 5xx 에러 발생 해결

seongjunnoh opened the pull request:
## #️⃣ 연관된 이슈

> closes #330 

## 📝 작업 내용
'모임방 참여 api' 에 비관적 락을 적용해, '인기 작가 모임방에 여러 유저들이 동시에 모임방 참여 요청을 보낼 경우' 를 재현한 테스트 코드와 부하 테스트에서의 5xx 에러 발생 이슈를 해결했습니다.

<주요 수정 부분>
- 최초 Room 엔티티 조회 시, select for update 패턴으로 x-lock 을 걸고 room 엔티티 조회하도록 코드 수정
- lock 대기 시간을 최대 5초로 설정 & 최대 3회 재시도 하도록 제한 (spring retry 활용)
- 재시도 횟수 초과한 요청에 대해서는 HTTP 423(Locked) error throw

<수정 결과>
- '모임방 참여' 에 대한 모든 쓰레드가 Room 엔티티의 x-lock 획득을 위해 대기하도록 하여 '동시성 이슈(= 데이터 정합성 안 맞음) 및 데드락으로 인한 5xx 에러 발생' 이슈를 해결
- But, x-lock 획득을 위한 대기 및 재시도로 인한 병목 발생으로 '서버의 요청 처리 시간' 이 길어지는 것 확인 (근거 : p(95) 지표, request duration 지표) 

## 📸 스크린샷
https://separate-snowplow-d8b.notion.site/API-2-2a3b701eb0b8809db7dbfafeda3596e0?source=copy_link
-> 정리한 노션 링크 입니다!

## 💬 리뷰 요구사항
- 서버의 처리율 개선을 위해, 모든 쓰레드가 Room 엔티티의 x-lock 획득을 위해 대기하는 구조를 개선하고자 합니다
- Room의 멤버 수를 update 하는 로직을 sql문으로 처리하면(-> 정원 초과 판단 & 멤버 수 update를 모두 하나의 sql문으로 처리), lock 없이 해결되겠지만, THIP 서버의 기반인 DDD 를 해친다고 판단했습니다
- 따라서 노션에 작성한 것처럼, '낙관적 락 기반의 Room 엔티티 update + 이벤트 기반의 RoomParticipant save' 구조로 모임방 참여 service 로직을 개선하려고 합니다
  - Redis를 활용해 '특정 room의 현재 멤버 수를 캐싱' 하는 방법도 생각했으나, '모임방 참여' 특성 상 Redis 자원 낭비 이슈가 더 크다고 판단했습니다

위 개선 방향에 대해 리뷰 달아주시면 감사하겠습니다!!

### 📌 PR 진행 시 이러한 점들을 참고해 주세요

    * P1 : 꼭 반영해 주세요 (Request Changes) - 이슈가 발생하거나 취약점이 발견되는 케이스 등
    * P2 : 반영을 적극적으로 고려해 주시면 좋을 것 같아요 (Comment)
    * P3 : 이런 방법도 있을 것 같아요~ 등의 사소한 의견입니다 (Chore)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

# 릴리스 노트

* **새로운 기능**
  * 방 참여 시 잠금 충돌에 대한 자동 재시도 도입으로 일시적 동시성 문제 완화

* **개선 사항**
  * 리소스가 잠긴 경우 명확한 오류 응답(자원 잠김) 제공
  * 참가자 중복을 방지하는 데이터 제약 추가로 방 참여 무결성 향상

* **테스트/운영**
  * 동시성 관련 부하 테스트 및 통계 집계 간소화로 모니터링 개선

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

seongjunnoh committed:
Merge branch 'test/#325-room-join-concurrency' into fix/#330-room-join-server-error

seongjunnoh committed:
[feat] x-lock 획득하여 Room 엔티티 조회하는 영속성 메서드 추가 (#330)

seongjunnoh committed:
[chore] spring retry 의존성 추가 (#330)

seongjunnoh committed:
[refactor] 비관적 락 + 재시도 로직을 적용하여 모임방 참여 service 코드 수정 (#330)

seongjunnoh committed:
[refactor] 모임방 참여 멀티쓰레드 테스트코드 수정 (#330)

- 데이터 정합성이 맞는 결과도 포함할 수 있도록 assert 코드 수정

seongjunnoh committed:
[refactor] 모임방 참여 k6 부하테스트 스크립트 수정 (#330)

- 모임방 참여 service 코드에 추가된 'RESOURCE_LOCKED exception throw' 를 커버하기 위해, HTTP 423 error도 캐치할 수 있도록 스크립트 수정

seongjunnoh committed:
[fix] 모임방 참여 service 단위 테스트 코드 수정 (#330)

- Room 엔티티 조회하는 메서드 변경으로 인한 테스트 코드 수정

seongjunnoh committed:
[fix] 모임방 참여 service 통합 테스트 코드 수정 (#330)

- 테스트 클래스의 트랜잭션 어노테이션 제거(수정된 모임방 참여 service 메서드가 매번 새로운 트랜잭션을 열기 떄문)
- MySQL, H2 DB의 락 획득 방식 차이로 인해 기존 테스트 코드 주석 처리
- test yml에 TIMEOUT 값 명시적으로 설정해도 테스트 코드 제대로 동작하지 X
- 일단 깨지는 테스트 메서드는 주석 처리
- 운영 환경과의 동일한 테스트를 위해 test DB를 MySQL로 변경하는게 좋아보임

seongjunnoh committed:
Merge remote-tracking branch 'origin' into fix/#330-room-join-server-error

seongjunnoh committed:
[refactor] 모임방 참여 service 멀티쓰레드 테스트 코드 수정 (#330)

- 멀티쓰레드 환경에서 동시에 모임방 참여 요청을 보낼 때, 200 OK 응답 수와 DB 데이터 정합성을 확인하는게 목적이라 4xx, 5xx error의 구분은 X

seongjunnoh commented on src/test/java/konkuk/thip/room/adapter/in/web/RoomJoinApiTest.java:
```diff
@@ -134,23 +155,28 @@ void joinRoom_success() throws Exception {
+//                .content(objectMapper.writeValueAsString(request)));
+//
+//        result.andExpect(status().isBadRequest());
+//    }
```
@hd0rable @buzz0331 커밋 메시지에 작성한 것처럼, 현재 브랜치에서 모임방 참여 코드의 동시성 이슈를 해결하기 위해 [비관적 락 + spring retry를 통한 재시도(매번 새로운 트랜잭션으로) + @Recover] 로직을 추가했는데, 기존 테스트 코드가 깨지는 부분 중 해결되지 않는 테스트 코드가 있는 상황입니다.

실패하는 테스트 메서드(= 4xx error 발생 확인을 위한 테스트) 의 빌드 로그를 확인하니, 트랜잭션 시작 이후 최초 Room 엔티티 조회 시 select for update sql 로그가 찍히고, 이후 @Recover로 흘러가 423 error 가 발생한 것을 확인했습니다.
(validation 과정까지 코드가 흘러가지 못해 400 error가 발생하지 않아서 테스트가 실패한 것임)

구글링을 통해 H2 DB는 제가 jpa repository에 명시한 @QueryHint의 timeout 설정값을 읽지 못한다는 정보를 얻어서, test yml의 H2 url에 LOCK_TIMEOUT 값을 명시했는데도 여전히 테스트 메서드가 깨졌습니다.
ref : https://rudaks.tistory.com/entry/Spring-Data-Jpa%EC%97%90%EC%84%9C-%EB%8D%B0%EC%9D%B4%ED%84%B0%EB%B2%A0%EC%9D%B4%EC%8A%A4%EB%B3%84-lock-timeout-%EC%84%A4%EC%A0%95-%ED%85%8C%EC%8A%A4%ED%8A%B8#google_vignette

일단 현재 브랜치의 모임방 참여 로직은 추후 개선될 예정이라, 깨지는 테스트 메서드를 일단 주석처리 하고 넘어가도 괜찮지 않나 라고 생각했는데, 다른 의견있으면 코멘트 부탁드립니다!!

추가로 현재 THIP이 실서비스 운영중이니, 서버의 테스트 메서드 환경의 DB을 운영 환경과 동일하게 MySQL로 변경하는 게 어떨지 제안드립니다!
(DB 환경이 달라서 마주하는 이슈들이 조금씩 크게 다가오는거 같습니다)
ref : https://dami97.tistory.com/73

hd0rable replied:
동시성이 있는 상황이아닌데 도대체 왜 423에러가 뜨는지궁금했었는데 단순 h2 스펙때문에 나는 오류였군요 확인했습니다!!
실제 운영 db와 테스트 db의환경이 달라서 저번에도 ci만을위해서 수정한 코드가 종종있었는데 Testcontainers라는 라이브러리가있었균요 좋은것같습니다!! 우선은 일단 주석처리하고 넘어가도 될 것같습니다

buzz0331 replied:
제 생각은 현재 `@Recover`와 `@Retryable`어노테이션으로 인해 원래 BusinessException이나 InvalidException으로 발생해야 하는 예외상황들이 모두 RESOURCE_LOCKED 에러로 치환되어서 throw되기 때문에 기존 테스트 케이스가 깨지는 것이 아닐까 라고 생각되는데 이부분은 어떻게 생각하시나요? (카톡에서 이야기한 내용과 일맥상통하는 이야기입니다!)

seongjunnoh commented on src/test/java/konkuk/thip/room/concurrency/RoomJoinConcurrencyTest.java:
```diff
@@ -132,11 +132,11 @@ void room_join_test_in_multi_thread() throws Exception {
         // 1) participants가 recruitCount 보다 커질 수 있음
-        assertThat(participantRows).isGreaterThan(recruit);
+        assertThat(participantRows).isGreaterThanOrEqualTo(recruit);
 
         // 2) memberCount가 실제 participants 수보다 작을 수 있음
         // memberCount 값은 Room 도메인 규칙에 의해 recruitCount를 초과하여 증가하지 않음
-        assertThat(memberCountInRoom).isLessThan((int) participantRows);
+        assertThat(memberCountInRoom).isLessThanOrEqualTo((int) participantRows);
```
@coderabbitai 현재 서비스 메서드에서 최초로 Room 엔티티를 조회할 때 x-lock을 획득하게 강제해서 동시성 이슈가 발생할 여지는 없지 않나?
그리고 커밋 메시지에 작성한 것처럼, 현재 테스트 메서드의 목적은 멀티쓰레드 환경으로 모임방 참여 요청이 동시에 몰리는 상황을 재현했을 때, 200 Ok 응답 개수와 DB에 기록된 데이터를 통해 데이터 정합성이 맞는지를 확인하기 위함이야
따라서 200 Ok가 아닌 다른 응답의 HTTP Status는 굳이 분류하지 않았어

hd0rable commented on src/main/java/konkuk/thip/room/application/service/RoomJoinService.java:
```diff
@@ -30,13 +35,21 @@ public class RoomJoinService implements RoomJoinUseCase {
+//                .orElseThrow(() -> new BusinessException(ErrorCode.USER_CANNOT_JOIN_OR_CANCEL));
+
+        /** 락 타잉아웃 발생 포인트 **/
+        Room room = roomCommandPort.getByIdForUpdate(roomJoinCommand.roomId());
```
p3: 존재하지 않는 방에대해 방참여/취소 요청에대한 예외를 기존에는 USER_CANNOT_JOIN_OR_CANCEL로 처리했었는데 수정된 코드에서는 ROOM_NOT_FOUND로 처리하고있는데 이렇게 수정하신 이유가 따로있나용??

seongjunnoh replied:
Room 엔티티 조회 시 x-lock을 걸고 조회하게끔 수정하는데 집중하다보니, 기존 코드처럼 조회가 되지 않을 경우에 대한 고려를 깊게 하지는 않았습니다!

현재 작업 중인 브랜치에서 이 부분 또한 고려해보도록 하겠습니다!

hd0rable commented on src/main/java/konkuk/thip/room/application/service/RoomJoinService.java:
```diff
@@ -30,13 +35,21 @@ public class RoomJoinService implements RoomJoinUseCase {
     @Override
-    @Transactional
+    @Retryable(
+            noRetryFor = {InvalidStateException.class, BusinessException.class},  // 재시도 제외 예외
```
좋네여

hd0rable reviewed (dismissed):
수고하셨습니닷!! 노션도 확인했습니다!! 단순궁금증 질문 남겨보았는데 확인 부탁드립니다!!

buzz0331 reviewed (dismissed):
너무 좋은 것 같습니다! 테스트 코드나 부하테스트 시나리오 의도가 명확해서 이해도 잘되고 잘 와닿는 것 같습니다~

seongjunnoh committed:
[refactor] 방 참여 api service 코드 수정 (#330)

- InvalidException, BusinessException에 대한 @Recover 메서드 추가
- @Retryable 메서드 내에서 발생한 exception은 모두 @Recover 설정해줘야함

seongjunnoh committed:
[refactor] 방 참여 API 통합 테스트 코드 수정 (#330)

- InvalidException, BusinessException에 대한 @Recover 메서드 추가로 인해 테스트 코드 주석 해제

seongjunnoh committed:
[refactor] 방 참여 API 부하 테스트 스크립트 수정 (#330)

- 테스트 수행 결과의 400, 423 응답 구분

seongjunnoh committed:
[chore] room_participants unique 제약 조건 추가 (#330)

- (user, room) 쌍에 대해 unique 제약 추가

seongjunnoh committed:
[fix] room_participants unique 제약 조건 추가에 따른 테스트 코드 수정 (#330)

seongjunnoh committed:
Merge remote-tracking branch 'origin' into fix/#330-room-join-server-error

seongjunnoh commented on src/main/java/konkuk/thip/room/application/service/RoomJoinService.java:
```diff
@@ -73,10 +81,20 @@ public RoomJoinResult changeJoinState(RoomJoinCommand roomJoinCommand) {
         throw new BusinessException(ErrorCode.RESOURCE_LOCKED);
     }
 
+    @Recover
+    public RoomJoinResult recoverInvalidStateException(InvalidStateException e, RoomJoinCommand roomJoinCommand) {
+        throw e;
+    }
+
+    @Recover
+    public RoomJoinResult recoverBusinessException(BusinessException e, RoomJoinCommand roomJoinCommand) {
+        throw e;
+    }
```
@buzz0331 @hd0rable 이전에 방 참여 API 실행 중 도메인 로직으로 인해 발생한 InvalidStateException, BusinessException 이 @Recover 메서드로 흡수되어 모두 423 error로 응답되었던 이슈를 해결했습니다

spring @Retryable 로 정의한 메서드내에서 발생하는 exception을 재시도 여부와 관계없이 각 exception에 대한 @Recover 메서드로 exception handling 이 필요하다고 하네요

seongjunnoh reviewed (commented):
@buzz0331 @hd0rable MySQL x-lock 적용하는 것으로 해당 pr merge 하고 조회 로직 개선으로 넘어가보겠습니다~~

현재 구현은 [DB x-lock + 락 타임아웃 발생 시 재시도] 로 동시에 여러 방 참여 요청이 들어오더라도 데이터 정합성을 유지할 수 있습니다.
[락 획득을 위한 대기시간 설정 + 타임아웃 시 재시도 로직]을 설정해 클라이언트가 [방 참여 성공 or API 제약 조건으로 인한 정상적인 방 참여 실패] 이렇게 2가지 응답만을 받을 수 있도록 했습니다.
-> 재시도 횟수가 초과됐음에도 락 획득에 실패한다면 423 error를 응답하도록 했습니다 (과도한 고트래픽이 아닌 이상 여기까지 오지는 않을 것으로 예상)

이때 잦은 재시도로 인해 톰캣 쓰레드의 점유 시간이 늘어나 서버 전체의 API 요청 처리량이 줄어드는 것을 방지하기 위해, 락 타임아웃 시간을 비교적 넉넉하게(= 5초) 로 설정했습니다.

하지만 현재 구현은 락 경합이 심할 경우 락 해제를 대기하는 쓰레드들의 DB 커넥션 점유 시간이 늘어나고, DB 커넥션 풀 고갈로 인해 방 참여 APi 뿐만아니라 다른 API 응답에서도 문제가 생길 수 있습니다.
-> 최대한 트랜잭션 점유 시간을 줄여 빠른 락 해제를 보장해야하는데, 이를 위한 방법은 고민이 필요할 것 같습니다

추후에 분산 환경을 고려한다면 DB에게 동시성 처리를 맡기는게 아니라, 어플리케이션단에서 redis 등을 활용한 분산 락을 적용하는 등 다른 해결책도 생각해보면 좋을 듯 합니다.

hd0rable reviewed (approved):
확인했습니다!! 수고하셨습니다~~

buzz0331 reviewed (approved):
좋네요. 확인했습니다~
